File: app/src/main/java/com/dmi/perfectreader/layout/liner/BreakLiner.java
package com.dmi.perfectreader.layout.liner;

import com.dmi.perfectreader.layout.config.LayoutChars;
import com.dmi.util.cache.ReuseCache.Reuser;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import static com.dmi.util.cache.ReuseCache.reuseCollection;
import static com.dmi.util.cache.ReuseCache.reuser;
import static com.google.common.base.Preconditions.checkState;
import static java.lang.Math.max;

public class BreakLiner implements Liner {
    private final BreakFinder breakFinder;

    public BreakLiner(BreakFinder breakFinder) {
        this.breakFinder = breakFinder;
    }

    @Override
    public List<Line> makeLines(MeasuredText measuredText, Config config) {
        List<Line> lines = Reusables.lines();

        CharSequence text = measuredText.plainText();
        Locale locale = measuredText.locale();

        new Runnable() {
            LinePart part = new LinePart();             // содержит символы, которые уже точно будет содержать строка
            LinePart newPart = new LinePart();          // содержит символы из part, а также новые добавочные. необходима для проверки, вмещается ли новая строка

            @Override
            public void run() {
                part.reset(0, 0, config.firstLineIndent(), false);

                breakFinder.findBreaks(text, locale, br -> {
                    checkState(part.endIndex < br.index(), "Wrong line end index");

                    pushChars(br.index(), br.hasHyphen());

                    if (br.isForce()) {
                        pushLine(true);
                    }

                    checkState(part.endIndex == br.index(), "Wrong line end index");
                });

                pushLine(true);
            }

            void pushChars(int endIndex, boolean hasHyphenAfter) {
                newPart.reset(part.beginIndex, endIndex, part.indent, hasHyphenAfter);

                if (canUsePart(newPart)) {
                    useNewPart();
                } else {
                    pushLine();
                    pushCharsAtBeginOfLine(endIndex, hasHyphenAfter);
                }
            }

            private void pushCharsAtBeginOfLine(int endIndex, boolean hasHyphenAfter) {
                newPart.reset(part.beginIndex, endIndex, part.indent, hasHyphenAfter);

                if (canUsePart(newPart)) {
                    useNewPart();
                } else {
                    pushCharByChar(endIndex, hasHyphenAfter);
                }
            }

            private void pushCharByChar(int endIndex, boolean hasHyphenAfter) {
                for (int end = part.endIndex + 1; end <= endIndex; end++) {
                    boolean charHasHyphenAfter = end == endIndex && hasHyphenAfter;
                    newPart.reset(part.beginIndex, end, part.indent, charHasHyphenAfter);

                    if (!canUsePart(newPart)) {
                        pushLine();
                        newPart.reset(part.beginIndex, end, part.indent, charHasHyphenAfter);
                    }

                    useNewPart();
                }
            }

            private boolean canUsePart(LinePart newPart) {
                return newPart.left + newPart.width <= config.maxWidth() ||
                       newPart.left + newPart.width == part.left + part.width;
            }

            private void useNewPart() {
                LinePart temp = part;
                part = newPart;
                newPart = temp;
            }

            void pushLine() {
                pushLine(false);
            }

            void pushLine(boolean isLast) {
                if (!part.isEmpty()) {
                    LineImpl line = new LineImpl();
                    line.left = part.left;
                    line.width = part.width;
                    line.hasHyphenAfter = part.hasHyphenAfter;
                    line.isLast = isLast;
                    addTokensInto(line.tokens, part.beginIndex, part.endIndex);
                    lines.add(line);

                    part.reset(part.endIndex, part.endIndex, 0, false);
                }
            }

            private void addTokensInto(List<Token> tokens, int beginIndex, int endIndex) {
                TokenImpl token = null;
                for (int begin = beginIndex; begin < endIndex; begin++) {
                    char ch = text.charAt(begin);
                    boolean isSpace = isSpace(ch);
                    int end = begin + 1;

                    if (token == null || token.isSpace() != isSpace) {
                        token = new TokenImpl();
                        token.isSpace = isSpace;
                        token.beginIndex = begin;
                        token.endIndex = end;
                        tokens.add(token);
                    } else {
                        token.endIndex = end;
                    }
                }
            }

            private boolean isSpace(char ch) {
                return Character.isSpaceChar(ch) || Character.isWhitespace(ch);
            }

            class LinePart {
                public int beginIndex;
                public int endIndex;
                public float indent;
                public boolean hasHyphenAfter;

                public float left;
                public float width;

                public void reset(int beginIndex, int endIndex, float indent, boolean hasHyphenAfter) {
                    this.beginIndex = beginIndex;
                    this.endIndex = endIndex;
                    this.indent = indent;
                    this.hasHyphenAfter = hasHyphenAfter;

                    if (endIndex > beginIndex) {
                        int trailingSpacesBeginIndex = findTrailingSpacesBeginIndex(beginIndex, endIndex);
                        float symbolsWidth = measuredText.widthOf(beginIndex, trailingSpacesBeginIndex);
                        float hyphenWidth = hasHyphenAfter ? measuredText.hyphenWidthAfter(endIndex - 1) : 0;

                        float leftHang = config.leftHangFactor(text.charAt(beginIndex)) * measuredText.widthOf(beginIndex);
                        float rightHang = rightHang(beginIndex, trailingSpacesBeginIndex, hyphenWidth, hasHyphenAfter);

                        left = indent - leftHang;
                        width = max(0, symbolsWidth + hyphenWidth - rightHang);
                    } else {
                        left = 0;
                        width = 0;
                    }
                }

                private float rightHang(int beginIndex, int trailingSpacesBeginIndex, float hyphenWidth, boolean hasHyphenAfter) {
                    int lastIndex = trailingSpacesBeginIndex - 1;
                    if (lastIndex >= beginIndex) {
                        char lastChar = text.charAt(lastIndex);
                        return hasHyphenAfter ?
                               config.rightHangFactor(LayoutChars.HYPHEN) * hyphenWidth :
                               config.rightHangFactor(lastChar) * measuredText.widthOf(lastIndex);
                    } else {
                        return 0;
                    }
                }

                private int findTrailingSpacesBeginIndex(int beginIndex, int endIndex) {
                    for (int i = endIndex - 1; i >= beginIndex; i--) {
                        char ch = text.charAt(i);
                        if (!isSpace(ch)) {
                            return i + 1;
                        }
                    }
                    return beginIndex;
                }

                public boolean isEmpty() {
                    return endIndex == beginIndex;
                }
            }
        }.run();

        return lines;
    }

    private static class Reusables {
        private static final Reuser<ArrayList<Line>> lines = reuser(ArrayList::new);

        public static ArrayList<Line> lines() {
            return reuseCollection(lines);
        }
    }

    private static class LineImpl implements Line {
        public float left;
        public float width;
        public boolean hasHyphenAfter;
        public boolean isLast;
        public List<Token> tokens = new ArrayList<>();

        @Override
        public float left() {
            return left;
        }

        @Override
        public float width() {
            return width;
        }

        @Override
        public boolean hasHyphenAfter() {
            return hasHyphenAfter;
        }

        @Override
        public boolean isLast() {
            return isLast;
        }

        @Override
        public List<Token> tokens() {
            return tokens;
        }
    }

    private static class TokenImpl implements Token {
        public boolean isSpace;
        public int beginIndex;
        public int endIndex;

        @Override
        public boolean isSpace() {
            return isSpace;
        }

        @Override
        public int beginIndex() {
            return beginIndex;
        }

        @Override
        public int endIndex() {
            return endIndex;
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/layout/liner/BreakLiner.kt
package com.dmi.perfectreader.layout.liner

import com.dmi.perfectreader.layout.config.LayoutChars
import com.dmi.perfectreader.layout.liner.Liner.Line
import com.dmi.perfectreader.layout.liner.Liner.Token
import com.dmi.perfectreader.layout.liner.breaker.Breaker
import com.dmi.util.cache.ReusableArrayList
import com.google.common.base.Preconditions.checkArgument
import java.lang.Math.max
import java.lang.Math.min

class BreakLiner(private val breaker: Breaker) : Liner {
    companion object {
        val PROCESS_CHARS_STEP = 4
    }

    override fun makeLines(measuredText: Liner.MeasuredText, config: Liner.Config): List<Liner.Line> {
        return object {
            val lines = Reusables.lines()
            val text = measuredText.plainText
            val locale = measuredText.locale
            val breaks = breaker.breakText(text, locale)

            fun makeLines(): List<Liner.Line> {
                var part = LinePart().apply { reset(0, config.firstLineIndent) }

                iteratePotentialEnds{ end ->
                    checkArgument(end > part.endIndex)

                    part.setEnd(end, false)
                    if (part.right > config.maxWidth) {
                        part = fitPartByBreak(part, part.beginIndex, part.endIndex)
                        pushLine(part)
                        part.reset(part.endIndex, 0F)
                    }

                    return@iteratePotentialEnds part.endIndex
                }

                if (!part.isEmpty)
                    pushLine(part)

                return lines
            }

            fun iteratePotentialEnds(action: (Int) -> Int) {
                if (text.length > 0) {
                    var end = min(PROCESS_CHARS_STEP, text.length)
                    while (end < text.length)
                        end = action(end) + PROCESS_CHARS_STEP

                    end = text.length
                    while (end <= text.length)
                        end = action(text.length) + 1
                }
            }

            fun fitPartByBreak(part: LinePart, beginIndex: Int, endIndex: Int): LinePart {
                var i = endIndex - 1
                while (i > beginIndex) {
                    if (breaks.hasBreakBefore(i)) {
                        part.setEnd(i, breaks.hasHyphenBefore(i))
                        if (part.right <= config.maxWidth)
                            return part
                    }
                    i--
                }

                return fitPartCharByChar(part, beginIndex, endIndex)
            }

            fun fitPartCharByChar(part: LinePart, beginIndex: Int, endIndex: Int): LinePart {
                var end = endIndex - 1
                while (end > beginIndex) {
                    part.setEnd(end, false)
                    if (part.right <= config.maxWidth)
                        return part
                    end--
                }

                return fitMinimalChars(part, beginIndex)
            }

            fun fitMinimalChars(part: LinePart, beginIndex: Int): LinePart {
                part.setEnd(beginIndex + 1, false)
                val minimalWidth = part.right

                var end = beginIndex + 2
                while (end <= text.length) {
                    part.setEnd(end, false)
                    if (part.right > minimalWidth)
                        break
                    end++
                }

                part.setEnd(end - 1, false)
                return part
            }

            fun pushLine(part: LinePart) {
                checkArgument(part.endIndex > part.beginIndex)
                lines.add(
                        Line().apply {
                            this.left = part.left
                            this.width = part.width
                            this.hasHyphenAfter = part.hasHyphenAfter
                            addTokensInto(tokens, part.beginIndex, part.endIndex)
                        }
                )
            }

            fun addTokensInto(tokens: MutableList<Liner.Token>, beginIndex: Int, endIndex: Int) {
                for (begin in beginIndex..endIndex - 1) {
                    val ch = text[begin]
                    val isSpace = isSpace(ch)
                    val end = begin + 1
                    val last = tokens.lastOrNull()

                    if (last == null || last.isSpace != isSpace) {
                        tokens.add(Token().apply {
                            this.isSpace = isSpace
                            this.beginIndex = begin
                            this.endIndex = end
                        })
                    } else {
                        last.endIndex = end
                    }
                }
            }

            fun isSpace(ch: Char) = Character.isSpaceChar(ch) || Character.isWhitespace(ch)

            inner class LinePart {
                var beginIndex = 0
                var endIndex = 0
                var indent = 0F
                var hasHyphenAfter = false

                var left = 0F
                var width = 0F

                val right: Float
                    get() = left + width
                val isEmpty: Boolean
                    get() = endIndex == beginIndex

                fun reset(initialIndex: Int, indent: Float) {
                    this.beginIndex = initialIndex
                    this.endIndex = initialIndex
                    this.indent = indent
                    this.hasHyphenAfter = false
                    this.left = indent
                    this.width = 0F
                }

                fun setEnd(endIndex: Int, hasHyphenAfter: Boolean) {
                    if (isEmpty) {
                        val leftHang = config.leftHangFactor(text[beginIndex]) * measuredText.widthOf(beginIndex)
                        left = indent - leftHang
                    }

                    this.endIndex = endIndex
                    this.hasHyphenAfter = hasHyphenAfter

                    val trailingSpacesBegin = trailingSpacesBegin(beginIndex, endIndex)
                    val symbolsWidth = measuredText.widthOf(beginIndex, trailingSpacesBegin)
                    val hyphenWidth = if (hasHyphenAfter) measuredText.hyphenWidthAfter(endIndex - 1) else 0F
                    val rightHang = rightHang(beginIndex, trailingSpacesBegin, hyphenWidth, hasHyphenAfter)
                    width = max(0F, symbolsWidth + hyphenWidth - rightHang)
                }

                private fun trailingSpacesBegin(beginIndex: Int, endIndex: Int): Int {
                    var i = endIndex - 1
                    while (i >= beginIndex) {
                        if (!isSpace(text[i]))
                            return i + 1
                        i--
                    }
                    return beginIndex
                }

                private fun rightHang(beginIndex: Int, trailingSpacesBegin: Int, hyphenWidth: Float, hasHyphenAfter: Boolean): Float {
                    val lastIndex = trailingSpacesBegin - 1
                    if (lastIndex >= beginIndex) {
                        if (hasHyphenAfter) {
                            return config.rightHangFactor(LayoutChars.HYPHEN) * hyphenWidth
                        } else {
                            val lastChar = text[lastIndex]
                            return config.rightHangFactor(lastChar) * measuredText.widthOf(lastIndex)
                        }
                    } else {
                        return 0F
                    }
                }
            }
        }.makeLines()
    }

    private object Reusables {
        val lines = ReusableArrayList<Liner.Line>()
    }
}
